pkg/utils/gardener/garden.go

```diff
@@ -92,22 +93,98 @@ func DomainIsDefaultDomain(domain string, defaultDomains []*Domain) *Domain {
 
 var gardenRoleReq = utils.MustNewRequirement(v1beta1constants.GardenRole, selection.Exists)
 
+// ReadGardenInternalDomain reads the internal domain information from the Garden cluster.
+func ReadGardenInternalDomain(
+	ctx context.Context,
+	c client.Reader,
+	namespace string,
+	enforceSecret bool,
+	seedDNSProvider *gardencorev1beta1.SeedDNSProviderConf,
+) (
+	*Domain,
+	error,
+) {
+	if seedDNSProvider != nil {
+		secret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{
+			Name:      seedDNSProvider.CredentialsRef.Name,
+			Namespace: seedDNSProvider.CredentialsRef.Namespace,
+		}}
+
+		if err := c.Get(ctx, client.ObjectKeyFromObject(secret), secret); err != nil {
+			return nil, fmt.Errorf("cannot fetch internal domain secret: %w", err)
+		}
+
+		return &Domain{
+			Domain:     seedDNSProvider.Domain,
+			Provider:   seedDNSProvider.Type,
+			Zone:       ptr.Deref(seedDNSProvider.Zone, ""),
+			SecretData: secret.Data,
+		}, nil
+	}
+
+	secret, err := ReadInternalDomainSecret(ctx, c, namespace, enforceSecret)
+	if err != nil {
+		return nil, err
+	}
+	if secret == nil && !enforceSecret {
+		return nil, nil
+	}
```

Comment:
I think it should be possible to shorten this:
```suggestion
	if err != nil || secret == nil && !enforceSecret {
		return nil, err
	}
```

Reply:
I think that this is more or less a personal preference. I find it easier to read when the statement is split into two. I applied the suggestion with https://github.com/gardener/gardener/pull/12663/commits/ba0ffd47bf415677c1fcfa071875281b2cabc6f2